mobile/src/main/java/com/example/wearlearn/SignUpActivity.java

```diff
@@ -0,0 +1,142 @@
+import butterknife.InjectView;
+
+public class SignUpActivity extends AppCompatActivity {
+    private static final String TAG = "SignupActivity";
+
+    @InjectView(R.id.input_name) EditText _nameText;
+    @InjectView(R.id.input_email) EditText _emailText;
+    @InjectView(R.id.input_password) EditText _passwordText;
+    @InjectView(R.id.btn_signup) Button _signupButton;
+    @InjectView(R.id.link_login) TextView _loginLink;
+    @InjectView(R.id.input_reEnterPassword) EditText _reEnterPasswordText;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_signup);
+        ButterKnife.inject(this);
+
+        _signupButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                signup();
+            }
+        });
+
+        _loginLink.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                // Finish the registration screen and return to the Login activity
+                Intent intent = new Intent(getApplicationContext(),LoginActivity.class);
+                startActivity(intent);
+                finish();
+            }
+        });
+    }
+
+    public void signup() {
+        Log.d(TAG, "Signup");
+
+        if (!validate()) {
+            onSignupFailed();
+            return;
+        }
+
+        _signupButton.setEnabled(false);
+
+        final ProgressDialog progressDialog = new ProgressDialog(SignUpActivity.this,
+                R.style.AppTheme);
+        progressDialog.setIndeterminate(true);
+        progressDialog.setMessage("Creating Account...");
+        progressDialog.show();
+
+        String name = _nameText.getText().toString();
+        String email = _emailText.getText().toString();
+        String password = _passwordText.getText().toString();
+        String reEnterPassword = _reEnterPasswordText.getText().toString();
+
+        // TODO: Implement signup logic here.
+
+        new android.os.Handler().postDelayed(
+                new Runnable() {
+                    public void run() {
+                        // On complete call either onSignupSuccess or onSignupFailed
+                        // depending on success
+                        onSignupSuccess();
+                        // onSignupFailed();
+                        progressDialog.dismiss();
+                    }
+                }, 3000);
+    }
+
+
+    public void onSignupSuccess() {
+        _signupButton.setEnabled(true);
+        Intent intent = new Intent(this, MenuActivity.class);
+        startActivity(intent);
+        setResult(RESULT_OK, null);
+        finish();
+    }
+
+    public void onSignupFailed() {
+        Toast.makeText(getBaseContext(), "Login failed", Toast.LENGTH_LONG).show();
+
+        _signupButton.setEnabled(true);
+    }
+
+    public boolean validate() {
```

Comment:
Metoda powinna być prywatna. 

Najlepiej by taki walidator został przeniesiony do osobnej klasy. Taka klasa mogłaby implementować interfejs 'Validator<T>' z metodą boolean validate(T inputs);

```
public interface Validator<T>{
    boolean validate(T inputs);
}


// tej klasie przekazać elementy widoku
public class SignUpInputs {
    //... 
}

public class SignUpInputsValidatorImpl implements Validator<SignUpInputs>{

    @Override
    public boolean validate(SignUpInputs inputs) {
        boolean result = false;

        //...

        return true;
    }
}
```

Takie podejście pozwala skutecznie odseparować logikę i podzielić odpowiedzialność między klasami, to pozwoli stworzyć testowalny kod. Przy takim podejściu można napisać testy samego validatora, później mockować go w testach związanych z aktywnością. 

To jeden z moich pomysłów, choć nie wiem jak on się ma do wzorców dla androida czy samej javy, znalazłem za to jeszcze takie rozwiązanie:
http://stackoverflow.com/a/11838715

Oczywiście nie musicie tego przerabiać, nawet nie namawiam, ale zachęcam, żebyście przemyśleli rozwiązania, które przytoczyłem.